docs(framework): Introduce migration guide for `Message API` (#5770)
adap/flower: framework/docs/source/how-to-upgrade-to-message-api.rst

API! This guide will walk you through the necessary steps to transition from Flower Apps
based on ``Strategy`` and ``NumPyClient`` to their equivalent using the new Message API.
This guide is relevant when updating pre-``1.21`` Flower apps to the latest stable
version.

Let's dive in!

Summary of changes
------------------

Thousands of Flower Apps have been created using the Strategies and |numpyclient_link|_
abstractions. With the introduction of the Message API, these apps can now take
advantage of a more powerful and flexible communication layer with the |message_link|_
abstraction being its cornerstone. Messages replace the previous `FitIns` and `FitRes`
data structures (and their equivalent for the other operations) into a single, unified
and more versatile data structure.

To fully take advantage of the new Message API, you will need to update your app's code
to use the new message-based communication patterns. This guide will show you how to:

1. Update your |serverapp_link|_ to make use of the new `Message`-based strategies. You
   won't need to use the `server_fn` anymore. The new strategies make it easier to

>    customize how the different FL rounds are executed, to retrieve results from your
>    run, and more.

Reviewer: ```suggestion
   customize how the different federated learning rounds are executed, to retrieve results from your
   run, and more.
```

2. Update your |clientapp_link|_ so it operates directly on `Message` objects received
   from the |serverapp_link|_. You will be able to keep most of the code from your
   |numpyclient_link|_ implementation but you won't need to create a new class anymore
   or use the helper `client_fn` function.

.. tip::

    The main payload Message objects carry are of type |recorddict_link|_. You can think
    of it as a dictionary that can hold other types of records, namely
    |arrayrecord_link|_, |metricrecord_link|_, and |configrecord_link|_. Please refer to

>     the documentation for each record for all the details on how they can be constructed

Reviewer: Instead of pointing to the documentation, we should do a mini-intro here as well. Or would this take too much room?

Author: done

    and adapted to your usecase. In this guide we won't delve into the specifics of each
    record type, but rather focus on the overall migration process.

Install update
--------------

The first step is to update the Flower version defined in the `pyproject.toml` in your
app:

.. code-block:: toml
    :caption: pyproject.toml
    :emphasize-lines: 2

    dependencies = [
        "flwr[simulation]>=1.21.0", # update Flower package
        # ...
    ]

Then, run the following command to install the updated dependencies:

.. code-block:: bash

    # Install the app with updated dependencies
    $ pip install -e .

> Update your ServerApp
> ---------------------

Reviewer: ```suggestion
Update your ``ServerApp``
-------------------------
```

|metricrecord_link|_ from federated stages and, optionally, metrics from evaluation
stages done at the `ServerApp`.

.. note::

    In addition to helper methods for working with PyTorch models, the
    |arrayrecord_link|_ class comes with a pair of methods to convert such record to and
    from a list of `NumPy` arrays (i.e. to ``to_numpy_ndarrays`` and
    ``from_numpy_ndarrays``). You may choose these methods if you aren't working with
    PyTorch models.

.. code-block:: python
    :emphasize-lines: 3,9,10,14,17,20

    from flwr.common import ArrayRecord, ConfigRecord, Context, MetricRecord
    from flwr.server import Grid, ServerApp
    from flwr.serverapp import FedAvg

    # Create ServerApp
    app = ServerApp()

    @app.main()
    def main(grid: Grid, context: Context) -> None:

        # Defined model to federate and extract parameters
        model = MyModel()
        arrays = ArrayRecord(global_model.state_dict())

        # Instantiate strategy
        strategy = FedAvg(fraction_train=0.5)

        # Start the strategy
        result = strategy.start(
            grid=grid,
            initial_arrays=arrays,
            num_rounds=3,

>         )

Reviewer: ```suggestion
        )
        print(result)
```

Update your ClientApp
---------------------

Similar to the `ServerApp`, the `ClientApp` no longer requires a helper function (i.e.
``client_fn`` ) that instantiates a |numpyclient_link|_ or base |client_link|_ object.
Instead, with the Message API, you get to define directly how the ClientApp operates on
`Message` objects received from the `ServerApp`.

> Remember `NumPyClient` came with two key built-in methods ``fit`` and ``evaluate`` that

Reviewer: ```suggestion
Remember `NumPyClient` came with two key built-in methods, ``fit`` and ``evaluate``, that
```

            # Return the evaluation results
            return float(loss), len(self.test_loader.dataset), {"accuracy": float(accuracy)}

    def client_fn(context: Context):
        # Return an instance of MyFlowerClient
        return MyFlowerClient().to_client()

    app = ClientApp(client_fn=client_fn)

Upgrading a ClientApp designed around the `NumPyClient` + `client_fn` abstractions to
the Message API would result in the following code. Note that the behavior of the
`ClientApp` is defined directly in its methods (i.e. a secondary class based on
`NumPyClient` is no longer needed).

The |clientapp_link|_ abstraction comes with built-in ``@app.train`` and
``@app.evaluate`` decorators. The arguments the associated methods receive have been
unified and they both operate on `Message` objects. Each method is responsible for
handling the incoming `Message` objects and returning the appropriate response (also as
a `Message`). Note that you'll still be able to use the functions you might have written
to, for example, train your model using the ML framework of your choice. In this example
those are represented by ``train_fn`` and ``test_fn``. Follow these steps to migrate

> your existing `Client App`:

Reviewer: ```suggestion
your existing ``ClientApp``:
```

    @app.evaluate()
    def evaluate(msg: Message, context: Context) -> Message:
        """Evaluate the model on local data."""

        # Identical to @app.train but returning only metrics
        # after doing local evaluation
        # ...

        # Do local evaluation
        loss, accuracy = test_fn(model, test_loader)

        # Construct reply Message
        # Retrun metrics relevant to usecase
        # THe weighting metric is also sent and will be used
        # to do weighted aggregation of metrics
        metrics = MetricRecord(
            {
                "eval_loss": loss,
                "eval_accuracy": accuracy,
                "num-examples": len(test_loader.dataset),
            }
        )
        # Construct RecorDict and add MetricRecord
        content = RecordDict({"metrics": metrics})
        return Message(content=content, reply_to=msg)

This concludes the migration guide!

.. tip::

    If you would like to create a new Flower App using the `Message API`, run the ``flwr
    new`` command and choose the appropriate template. Alternatively, you may want to
    take a look at the `quickstart-pytorch

>     <https://github.com/adap/flower/blob/main/examples/quickstart-pytorch>`_ example.

Reviewer: Let's move this tip to the top. It'll be easier for users to migrate their existing apps if they have a fully working projects as a starting point.

    record type, but rather focus on the overall migration process.

Install update
--------------

The first step is to update the Flower version defined in the `pyproject.toml` in your
app:

.. code-block:: toml
    :caption: pyproject.toml
    :emphasize-lines: 2

    dependencies = [
        "flwr[simulation]>=1.21.0", # update Flower package
        # ...
    ]

Then, run the following command to install the updated dependencies:

.. code-block:: bash

    # Install the app with updated dependencies
    $ pip install -e .

Update your ``ServerApp``
-------------------------

Starting with Flower 1.21, the `ServerApp` no longer requires a `server_fn` function to
make use of strategies. This is because a new collection of strategies (all sharing the
common |strategy_link|_ base class) has been created to operate directly on `Message`
objects, allowing for a more streamlined and flexible approach to federated learning
rounds.

.. note::

    The new `Message`-based strategies are located in the `flwr.serverapp
    <ref-api/flwr.serverapp.html>`_ module unlike the previous strategies which were
    located in the `flwr.server.strategy <ref-api/flwr.server.strategy.html>`_ module.

>     Over time more strategies will be added to the `flwr.serverapp` module. Users are

Reviewer 2: ```suggestion
    Over time more strategies will be added to the `flwr.serverapp.strategy` module. Users are
```
